crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2027,9 +2027,20 @@ impl<'db> TypeInferenceBuilder<'db> {
         // TODO: proper typed call signature, representing keyword args etc
         let arg_types = self.infer_arguments(arguments);
         let function_type = self.infer_expression(func);
-        function_type
-            .call(self.db, arg_types.as_slice())
-            .unwrap_with_diagnostic(self.db, func.as_ref().into(), self)
+
+        // If the function is a bool (for example `x = bool(1)`), we can do some special handling
+        // and return more specific types.
+        return if function_type == builtins_symbol_ty(self.db, "bool") {
+            self.infer_bool_call_expression(arg_types.as_slice())
```

Comment:
This should happen inside `Type::call` instead of here, and it should use `FunctionType::is_stdlib_symbol` instead of querying the builtin type and comparing for equality.